File: app/src/main/java/com/github/ilikeyourhat/sudokusolver/solving/deduction/solver/DeductionSolverV3.java
package com.github.ilikeyourhat.sudokusolver.solving.deduction.solver;

import com.github.ilikeyourhat.sudokusolver.solving.deduction.algorithm.*;
import com.github.ilikeyourhat.sudokusolver.model.SudokuType;

import java.util.ArrayList;
import java.util.List;

public class DeductionSolverV3 extends DeductionSolver {

    @Override
    protected List<DeductionAlgorithm.Factory> provideAlgorithms(SudokuType type) {
        List<DeductionAlgorithm.Factory> algorithms = new ArrayList<>();
        algorithms.add(new HintEliminationAlgorithm.Factory());
        algorithms.add(new NakedSinglesAlgorithm.Factory());
        algorithms.add(new HiddenSinglesAlgorithm.Factory());
        for (int i = 2; i <= type.getRegionSize() / 2; i++) {
            algorithms.add(new NakedValuesAlgorithm.Factory(i));
            algorithms.add(new HiddenValuesAlgorithm.Factory(i));
        }
        algorithms.add(new RegionIntersectionAlgorithm.Factory());
        return algorithms;
    }
}

File: app/src/main/kotlin/com/github/ilikeyourhat/kudoku/solving/deduction/solver/DeductionSolverV3.kt
package com.github.ilikeyourhat.kudoku.solving.deduction.solver

import com.github.ilikeyourhat.kudoku.model.SudokuType
import com.github.ilikeyourhat.kudoku.solving.deduction.algorithm.DeductionAlgorithm
import com.github.ilikeyourhat.kudoku.solving.deduction.algorithm.HiddenSinglesAlgorithm
import com.github.ilikeyourhat.kudoku.solving.deduction.algorithm.HiddenValuesAlgorithm
import com.github.ilikeyourhat.kudoku.solving.deduction.algorithm.HintEliminationAlgorithm
import com.github.ilikeyourhat.kudoku.solving.deduction.algorithm.NakedSinglesAlgorithm
import com.github.ilikeyourhat.kudoku.solving.deduction.algorithm.NakedValuesAlgorithm
import com.github.ilikeyourhat.kudoku.solving.deduction.algorithm.RegionIntersectionAlgorithm

class DeductionSolverV3 : DeductionSolver() {

    override fun provideAlgorithms(type: SudokuType): List<DeductionAlgorithm.Factory> {
        val algorithms = mutableListOf(
            HintEliminationAlgorithm.Factory(),
            NakedSinglesAlgorithm.Factory(),
            HiddenSinglesAlgorithm.Factory()
        )
        for (i in 2..type.possibleValues / 2) {
            algorithms.add(NakedValuesAlgorithm.Factory(i))
            algorithms.add(HiddenValuesAlgorithm.Factory(i))
        }
        algorithms.add(RegionIntersectionAlgorithm.Factory())
        return algorithms
    }
}
